pkg/services/tree/sync.go

```diff
@@ -0,0 +1,98 @@
+package tree
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama"
+	"github.com/nspcc-dev/neofs-node/pkg/network"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"google.golang.org/grpc"
+)
+
+// Synchronize tries to synchronize log starting from the last stored height.
+func (s *Service) Synchronize(ctx context.Context, cid *cid.ID, treeID string) error {
+	nodes, err := s.getContainerNodes(cid)
+	if err != nil {
+		return fmt.Errorf("can't get container nodes: %w", err)
+	}
+
+	lm, err := s.forest.TreeGetOpLog(cid, treeID, 0)
+	if err != nil && !errors.Is(err, pilorama.ErrTreeNotFound) {
+		return err
+	}
+
+	height := lm.Time + 1
+	for _, n := range nodes.Flatten() {
+		n.IterateAddresses(func(addr string) bool {
+			var a network.Address
+			if err := a.FromString(addr); err != nil {
+				return false
+			}
+
+			cc, err := grpc.DialContext(ctx, a.HostAddr(), grpc.WithInsecure())
+			if err != nil {
+				// Failed to connect, try the next address.
+				return false
+			}
+
+			treeClient := NewTreeServiceClient(cc)
+			for {
+				h, err := s.synchronizeSingle(ctx, cid, treeID, height, treeClient)
+				if err != nil || h <= height {
+					// Error with the response, try the next node.
+					return true
+				}
+				height = h
+			}
+		})
+	}
+	return nil
+}
+
+func (s *Service) synchronizeSingle(ctx context.Context, cid *cid.ID, treeID string, height uint64, treeClient TreeServiceClient) (uint64, error) {
+	for {
+		req := &GetOpLogRequest{
+			Body: &GetOpLogRequest_Body{
+				ContainerId: cid.ToV2().GetValue(),
+				TreeId:      treeID,
+				Height:      height,
+				Count:       MaxGetSubTreeDepth,
+			},
+		}
+		if err := signMessage(req, s.key); err != nil {
+			return 0, err
+		}
+
+		c, err := treeClient.GetOpLog(ctx, req)
+		if err != nil {
+			return 0, fmt.Errorf("can't initialize client: %w", err)
+		}
+
+		res, err := c.Recv()
+		for ; err == nil; res, err = c.Recv() {
+			lm := res.GetBody().GetOperation()
+			m := &pilorama.Move{
+				Parent: lm.ParentId,
+				Child:  lm.ChildId,
+			}
+			if err := m.Meta.FromBytes(lm.Meta); err != nil {
+				return 0, err
+			}
+			if err := s.forest.TreeApply(cid, treeID, m); err != nil {
+				return 0, err
+			}
+			if m.Time > height {
+				height = m.Time + 1
+			} else {
+				height += 1
+			}
+		}
+		if !errors.Is(err, io.EOF) {
+			return 0, err
```

Comment:
Here and above, by returning zero, we seem to discard all potential progress. Is it possible to continue from recent `height` value in case of a break?

Reply:
Yes, fixed.